ParabolInc/parabol
# Org backup

Resolve #5043

This PR is only scoped to the dumping of data, since fetch/restore/documentation depends upon this funtionality.

 - [ ] backupOrganization creates a DB dump for the schema, which includes all data definitions including tables, FK constraints, indexes, sequence tables, etc.
 - [ ] backupOrganization creates a DB dump for the data rows that is filtered to only contain rows that pertain to the provided orgId. If possible, this dump happens directly without cloning the rows to a separate DB.
 - [ ] dumps all pg tables

Testing instructions:
  - go to admin graphiql page locally, click on private schema, do "backupOrganization" mutation
  - test with a locally created organization, as this is the current best way to ensure all pg data exists locally
  - the mutation will timeout in graphiql, and will return ```{
  "message": "Unexpected end of JSON input",
  "stack": "SyntaxError: Unexpected end of JSON input\n    at fetcher (webpack-internal:///./packages/client/modules/admin/containers/Graphql/GraphqlContainer.tsx:74:16)"
}```. this is normal, and it's what it's always done even before adding the postgres part.
  - both `schemaDump.tar.gz` and `orgBackupData.sql` should appear in the root of the project dir under `artifacts` folder
  - just look under created "orgBackup" schema in pgAdmin to verify rows have been dumped relevant to the org. this data along with schema will be the content of the dumps.
  - let's skip manually restoring the backups in the test steps, as this borders on testing `pg_dump` command itself. there are a couple edge cases involved in restoring which will be fully detailed in the documentation. 
  
Organization requirements:
  - the locally created organization should be github integrated (GitHubAuth table)
  - the locally created organization should have sprint poker activity with jira integrated objects and pushed estimate fields (AtlassianAuth and TaskEstimate table)
  - organization should have its own custom template (TemplateRef and TemplateScaleRef tables)
  - organization should have discussion threads (Discussion table)
  - for Poll and PollOption, there is no UI yet for them so just create them manually in pgAdmin, making sure to pass FK constraints.

Restore instructions:
```
cd <:path_to_parabol>/artifacts
```
```
# this will wipe all pg data and start from a clean slate
pg_restore --dbname=postgresql://pgparaboladmin:parabol@localhost:5432/parabol-saas schemaDump.tar.gz --clean
```
```
psql postgresql://pgparaboladmin:parabol@localhost:5432/parabol-saas -c 'ALTER SCHEMA public RENAME TO "orgBackup";'
```
```
pg_restore --dbname=postgresql://pgparaboladmin:parabol@localhost:5432/parabol-saas orgBackupData.tar.gz --disable-triggers
```
```
psql postgresql://pgparaboladmin:parabol@localhost:5432/parabol-saas -c 'ALTER SCHEMA "orgBackup" RENAME TO "public";'
```
